src/rapids_singlecell/preprocessing/_neighbors.py

```diff
@@ -356,3 +362,176 @@ def neighbors(
+        **distances** : sparse matrix of dtype `float32`.
+            Instead of decaying weights, this stores distances for each pair of
+            neighbors.
+    """
+
+    adata = adata.copy() if copy else adata
+    if adata.is_view:
+        adata._init_as_actual(adata.copy())
+    X = _choose_representation(adata, use_rep=use_rep, n_pcs=n_pcs)
+
+    X_contiguous = _check_neighbors_X(X, algorithm)
+    _check_metrics(algorithm, metric)
+
+    n_neighbors = neighbors_within_batch
+
+    if batch_key is None:
+        raise ValueError("Please provide a batch key to perform batch-balanced KNN.")
+
+    if batch_key not in adata.obs:
+        raise ValueError(f"Batch key '{batch_key}' not present in `adata.obs`.")
+
+    n_obs = adata.shape[0]
+    batch_list = adata.obs[batch_key].values
+    batches = np.unique(batch_list)
+    total_neighbors = n_neighbors * len(batches)
+    knn_dist = cp.zeros((X.shape[0], total_neighbors), dtype=np.float32)
+    knn_indices = cp.zeros_like(knn_dist).astype(int)
+
+    for to_ind in range(len(batches)):
+        batch_to = batches[to_ind]
+        mask_to = batch_list == batch_to
+        X_to = X_contiguous[mask_to]
+        ind_to = cp.arange(len(batch_list))[mask_to]
+
+        if algorithm == "brute":
+            sub_ind, sub_dist = _brute_knn(
+                X_to, X_contiguous, n_neighbors, metric, metric_kwds
+            )
+        elif algorithm == "cagra":
+            sub_ind, sub_dist = _cagra_knn(X_to, X_contiguous, n_neighbors, metric)
+        elif algorithm == "ivfflat":
+            sub_ind, sub_dist = _ivf_flat_knn(X_to, X_contiguous, n_neighbors, metric)
+        elif algorithm == "ivfpq":
+            sub_ind, sub_dist = _ivf_pq_knn(X_to, X_contiguous, n_neighbors, metric)
+        col_range = cp.arange(to_ind * n_neighbors, (to_ind + 1) * n_neighbors)
+        knn_indices[:, col_range] = ind_to[sub_ind]
+        knn_dist[:, col_range] = sub_dist
+
+    n_nonzero = n_obs * total_neighbors
+    rowptr = cp.arange(0, n_nonzero + 1, total_neighbors)
+    distances = cp_sparse.csr_matrix(
+        (cp.ravel(knn_dist), cp.ravel(knn_indices), rowptr), shape=(n_obs, n_obs)
+    )
+
+    set_op_mix_ratio = 1.0
+    local_connectivity = 1.0
+    X_conn = cp.empty((n_obs, 1), dtype=np.float32)
+    connectivities = fuzzy_simplicial_set(
+        X_conn,
+        n_neighbors,
+        random_state,
+        metric=metric,
+        knn_indices=knn_indices,
+        knn_dists=knn_dist,
+        set_op_mix_ratio=set_op_mix_ratio,
+        local_connectivity=local_connectivity,
+    )
+    connectivities = connectivities.tocsr().get()
+    distances = distances.get()
+    if key_added is None:
+        key_added = "neighbors"
+        conns_key = "connectivities"
+        dists_key = "distances"
+    else:
+        conns_key = key_added + "_connectivities"
+        dists_key = key_added + "_distances"
+    adata.uns[key_added] = {}
+
+    neighbors_dict = adata.uns[key_added]
```

Comment:
Maybe just make the dict first, and then add it into `adata` instead of inserting empty, taking it out, and then modifying? (see phil's comment just below this)